contracts/SeriesIndexer.sol

```diff
@@ -0,0 +1,792 @@
+        return seriesDetails[seriesID].metadataURI;
+    }
+
+    /**
+     * @notice Returns contract-level token data (e.g., IPFS CID) of a given series
+     */
+    function getSeriesContractTokenDataURI(uint256 seriesID) external view returns (string memory) {
+        return seriesDetails[seriesID].contractTokenDataURI;
+    }
+
+    /**
+     * @notice Returns the pending co-artist requests for a given artist address
+     */
+    function getArtistPendingCoArtistSeries(address artistAddr) external view returns (uint256[] memory) {
+        uint256 artistID = addressToArtistID[artistAddr];
+        return artists[artistID].pendingCoArtistSeries;
+    }
+
+    /**
+     * @notice Returns the pending co-artist list for a given series
+     */
+    function getSeriesPendingCoArtists(uint256 seriesID) external view returns (address[] memory) {
+        uint256[] memory artistIDs = seriesDetails[seriesID].pendingCoArtists;
+        address[] memory artistAddrs = new address[](artistIDs.length);
+        for (uint256 i = 0; i < artistIDs.length; i++) {
+            artistAddrs[i] = artists[artistIDs[i]].artistAddress;
+        }
+        return artistAddrs;
+    }
+
+    /**
+     * @notice Returns the address associated with a given artist ID
+     */
+    function getArtistAddress(uint256 artistID) external view returns (address) {
+        return artists[artistID].artistAddress;
+    }
+
+    /**
+     * @notice Returns the artist ID associated with a given address
+     */
+    function getAddressArtistID(address artistAddr) external view returns (uint256) {
+        return addressToArtistID[artistAddr];
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @dev Creates a new series with the specified artists
+     */
+    function _createSeries(
+        address[] memory artistAddrs,
+        string memory metadataURI,
+        string memory tokenIDsMapURI
+    ) internal returns (uint256) {
+        if (artistAddrs.length == 0) {
+            revert NoArtistsForSeriesError();
+        }
+        if (msg.sender == owner()) {
+            _validateArtistsNotRevoked(artistAddrs);
+        } else {
+            _validateOnlySelfAsArtist(artistAddrs);
+        }
+        uint256 seriesID = nextSeriesID++;
+        _validateMetadataAndTokenURI(metadataURI, tokenIDsMapURI);
+
+        Series storage series = seriesDetails[seriesID];
+        series.metadataURI = metadataURI;
+        series.contractTokenDataURI = tokenIDsMapURI;
+
+        uint256[] memory artistIDs = _addArtistsToSeries(seriesID, artistAddrs);
+
+        emit SeriesIndexed(seriesID, artistIDs, metadataURI, tokenIDsMapURI);
+        return seriesID;
+    }
+
+    /**
+     * @dev Update existing series' metadata and token data
+     */
+    function _updateSeries(
+        uint256 seriesID,
+        string memory metadataURI,
+        string memory tokenIDsMapURI
+    ) internal seriesExists(seriesID) onlyOwnerOrArtist(seriesID) {
```

Comment:
I'd preferring using the modifiers in public/external function to leverage the visibility of them, we also could prevent potential gas waste when calling internal function multiple times.